### PR TITLE
check memory limits for cgroupv1

### PR DESCRIPTION
**Pull request by simonbyrne**
Seems to fix issue with https://github.com/JuliaLang/julia/pull/46796#issuecomment-1272117544

I admit I know very little about cgroups, or C programming, so this may well be incorrect, but it was able to correctly determine the cgroup limits on our Slurm cluster

### CODE REVIEW

File: src/unix/linux-core.c [hunk 1]
```diff
@@ -811,29 +811,48 @@ uint64_t uv_get_constrained_memory(void) {
...
+    p = buf + 4;
+
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.max", p);
+    max = uv__read_uint64(filename);
+
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.high", p);
+    high = uv__read_uint64(filename);
+  } else {
+    p = strchr(buf, ':');
+    while (p != NULL && memcmp(p, ":memory:", 8)) {
```
Comment (simonbyrne) [inline src/unix/linux-core.c:834]: oh wait, I realise this isn't how cgroups work...

**Comment by maleadt** [comment]
The fallbacks were already intended to address cgroupv1. What is wrong with those?

**Comment by simonbyrne** [comment]
It isn't giving the correct result on our Slurm cluster. If I do:
```
├ srun -t 01:10:00 --mem 2G --pty bash -l # start job with 2G of memory

├ cat /sys/fs/cgroup/memory/memory.limit_in_bytes # what is read by uv__get_constrained_memory_fallback
9223372036854771712

├ cat /proc/self/cgroup # cgroups of current process
11:blkio:/
10:memory:/slurm/uid_5184/job_28925944/step_0
9:freezer:/slurm/uid_5184/job_28925944/step_0
8:pids:/
7:cpuset:/slurm/uid_5184/job_28925944/step_0
6:net_prio,net_cls:/
5:perf_event:/
4:hugetlb:/
3:devices:/slurm/uid_5184/job_28925944/step_0/task_0
2:cpuacct,cpu:/
1:name=systemd:/system.slice/slurmd.service

├ cat /sys/fs/cgroup/memory/slurm/uid_5184/job_28925944/step_0/memory.limit_in_bytes # current memory controller cgroup
2147483648
```

File: src/unix/linux-core.c (hunk 1 above)
Comment (simonbyrne) [reply src/unix/linux-core.c:834]: Actually, I think it is correct? But it would be good for someone who understands these things to check.

**Comment by simonbyrne** [comment]
Any thoughts on this? I'd really like to get this into 1.9 (or ideally backported to 1.8).

File: src/unix/linux-core.c (hunk 1 above)
Comment (maleadt) [reply src/unix/linux-core.c:834]: I take it slurm creates a cgroup subtree? On my cgroupv1 system, the limit sits at `/sys/fs/cgroup/memory` directly. It'd be useful to add a comment here about what you're doing, that's easier to review than to parse C code 🙂

File: src/unix/linux-core.c [hunk 2]
```diff
@@ -811,29 +811,48 @@ uint64_t uv_get_constrained_memory(void) {
...
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/memory.limit_in_bytes", p);
+    max = uv__read_uint64(filename);
 
-  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.max", p);
-  max = uv__read_uint64(filename);
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/memory.soft_limit_in_bytes", p);
+    high = uv__read_uint64(filename);
+  }
 
   if (max == 0)
```
Comment (maleadt) [inline src/unix/linux-core.c:853]: `if (max == 0 || high == 0)`

**Review by maleadt — COMMENTED** [review]
LGTM. I'm sure upstream will have more comments though, so could you create a PR there once I have our pending changes merged?

**Comment by maleadt** [comment]
Actually, I'll try to upstream it, since I'm working on that right now anyway.

File: src/unix/linux-core.c (hunk 1 above)
Comment (simonbyrne) [reply src/unix/linux-core.c:834]: Yes, that appears to be how cgroupv1 works, e.g. https://unix.stackexchange.com/a/621576

**Comment by simonbyrne** [comment]
> Actually, I'll try to upstream it, since I'm working on that right now anyway.

Thanks, please let me know if I can do anything to help.

File: src/unix/linux-core.c (hunk 1 above)
Comment (maleadt) [reply src/unix/linux-core.c:834]: Apparently we're still missing something though, as on my cgroup1 system with Docker, I can't access even the memory subgroup listed in `/proc/self/cgroup`.

**Comment by maleadt** [comment]
I updated your PR to still correctly read out the limits I'm seeing (ref https://github.com/JuliaLang/libuv/pull/27#discussion_r993128302) and pushed that to https://github.com/libuv/libuv/pull/3754. Can you verify that works with your SLURM system?

**Comment by maleadt** [comment]
Superseded by https://github.com/JuliaLang/libuv/pull/32

**Pull request closed without merging**